core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/StackOperationService.java

```diff
@@ -113,54 +113,57 @@ public FlowIdentifier removeInstance(Stack stack, String instanceId, boolean for
-                        updateNodeCountValidator.validateScalingAdjustment(instanceGroupName, scalingAdjustment, stack);
-                    }
-                }
-                if (scalingStrategy != null) {
-                    // TODO: CB-14929: CB-15154 stopstart can only support a single hostGroup being specified.
-                    if (instanceIdsByHostgroupMap.size() > 1) {
-                        throw new BadRequestException("Downscale via Instance Stop cannot process more than one host group");
-                    }
-                    stackUpdater.updateStackStatus(stack.getId(), DetailedStackStatus.DOWNSCALE_BY_STOP_REQUESTED,
-                            "Requested node count for downscaling (stopstart): " + instanceIds.size());
-                    return flowManager.triggerStopStartStackDownscale(stack.getId(), instanceIdsByHostgroupMap, forced);
-                } else {
-                    return flowManager.triggerStackRemoveInstances(stack.getId(), instanceIdsByHostgroupMap, forced);
-                }
-            });
-        } catch (TransactionExecutionException e) {
-            if (e.getCause() instanceof BadRequestException) {
-                throw e.getCause();
+    public FlowIdentifier stopInstances(Stack stack, Collection<String> instanceIds, boolean forced) {
+        LOGGER.info("Received stop instances request for instanceIds: [{}]", instanceIds);
+
+        if (instanceIds == null || instanceIds.isEmpty()) {
+            throw new BadRequestException("Stop request cannot process an empty instanceIds collection");
+        }
+        Map<String, Set<Long>> instanceIdsByHostgroupMap = new HashMap<>();
+        for (String instanceId : instanceIds) {
+            InstanceMetaData metaData = updateNodeCountValidator.validateInstanceForStop(instanceId, stack);
+            if (metaData != null) {
+                instanceIdsByHostgroupMap.computeIfAbsent(metaData.getInstanceGroupName(), s -> new LinkedHashSet<>()).add(metaData.getPrivateId());
             }
-            throw new TransactionRuntimeExecutionException(e);
         }
+        if (instanceIdsByHostgroupMap.size() > 1) {
+            throw new BadRequestException("Downscale via Instance Stop cannot process more than one host group");
+        }
+        updateNodeCountValidator.validateServiceRoles(stack, instanceIdsByHostgroupMap.entrySet()
+                .stream()
+                .collect(Collectors.toMap(Entry::getKey, e -> e.getValue().size() * -1)));
+        LOGGER.info("Stopping the following instances: {}", instanceIdsByHostgroupMap);
+        if (!forced) {
+            for (Entry<String, Set<Long>> entry : instanceIdsByHostgroupMap.entrySet()) {
+                String instanceGroupName = entry.getKey();
+                int scalingAdjustment = entry.getValue().size() * -1;
+                updateNodeCountValidator.validateScalabilityOfInstanceGroup(stack, instanceGroupName, scalingAdjustment);
```

Comment:
This, and the next validation make use of hostGroupNodeCount - which does not necessarily reflect the state of the services running on the nodes. For the downscale scenario - this may be OK. Would be better to have exact validations though (i.e. consider the state of the nodes instead of the node counts).

Reply:
validateScalabilityOfInstanceGroup also checks for black-listed roles, that is why it still shows up in the code.